integration/container_event_test.go

```diff
@@ -0,0 +1,133 @@
+	assert.NoError(t, err)
+	containerEventsChan := make(chan *runtime.ContainerEventResponse)
+
+	go listenToEventChannel(ctx, t, containerEventsChan, containerEventsStreamingClient)
+	// Sleep to make sure the event channel is set up, and then drain all events
+	// emitted by previous tests.
+	time.Sleep(2 * time.Second)
+	drainContainerEventsChan(t, containerEventsChan)
+
+	t.Logf("Create a pod config and run sandbox container")
+	sandboxName := "container_events_sandbox"
+	sb, sbConfig := PodSandboxConfigWithCleanup(t, sandboxName, "container_events")
+
+	assert.NoError(t, checkContainerEventResponse(t, containerEventsChan, runtime.ContainerEventType_CONTAINER_CREATED_EVENT, sandboxName, ""))
+	assert.NoError(t, checkContainerEventResponse(t, containerEventsChan, runtime.ContainerEventType_CONTAINER_STARTED_EVENT, sandboxName, ""))
+
+	pauseImage := images.Get(images.Pause)
+	EnsureImageExists(t, pauseImage)
+	t.Logf("Create a container config and run container in a pod")
+	containerConfig := ContainerConfig(
+		"container1",
+		pauseImage,
+		WithTestLabels(),
+		WithTestAnnotations(),
+	)
+	cn, err := runtimeService.CreateContainer(sb, containerConfig, sbConfig)
+	require.NoError(t, err)
+	checkContainerEventResponse(t, containerEventsChan, runtime.ContainerEventType_CONTAINER_CREATED_EVENT, sandboxName, cn)
+	defer func() {
+		assert.NoError(t, runtimeService.RemoveContainer(cn))
+		assert.NoError(t, checkContainerEventResponse(t, containerEventsChan, runtime.ContainerEventType_CONTAINER_DELETED_EVENT, sandboxName, cn))
+	}()
+
+	require.NoError(t, runtimeService.StartContainer(cn))
+	assert.NoError(t, checkContainerEventResponse(t, containerEventsChan, runtime.ContainerEventType_CONTAINER_STARTED_EVENT, sandboxName, cn))
+	defer func() {
+		assert.NoError(t, runtimeService.StopContainer(cn, 10))
+		assert.NoError(t, checkContainerEventResponse(t, containerEventsChan, runtime.ContainerEventType_CONTAINER_STOPPED_EVENT, sandboxName, cn))
+	}()
+}
+
+func listenToEventChannel(ctx context.Context, t *testing.T, containerEventsChan chan *runtime.ContainerEventResponse, containerEventsStreamingClient runtime.RuntimeService_GetContainerEventsClient) {
+	t.Helper()
+	for {
+		resp, err := containerEventsStreamingClient.Recv()
+		if err == io.EOF {
+			return
+		}
+		assert.NoError(t, err)
+		if resp != nil {
+			containerEventsChan <- resp
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+	}
+}
+
+func drainContainerEventsChan(t *testing.T, containerEventsChan chan *runtime.ContainerEventResponse) {
+	for {
+		select {
+		case <-containerEventsChan:
+		case <-time.After(500 * time.Millisecond):
+			return
+		}
+	}
+}
+
+func checkContainerEventResponse(t *testing.T, containerEventsChan chan *runtime.ContainerEventResponse, expectedType runtime.ContainerEventType, expectedSandboxName string, expectedContainerID string) error {
+	t.Helper()
```

Comment:
nit: no need to pass `*testing.T` or call `t.Helper()` since no other methods of `*testing.T` are used

```suggestion
func checkContainerEventResponse(containerEventsChan chan *runtime.ContainerEventResponse, expectedType runtime.ContainerEventType, expectedSandboxName string, expectedContainerID string) error {
```

Reply:
good catch. I added it for logging when debugging the test failure but forgot to clean it up.

Reply:
Got it; you can leave that in there (it's useful too).